apps/argus/src/services/controller_service.rs

```diff
@@ -125,3 +161,595 @@ impl Service for ControllerService {
         Ok(())
     }
 }
+
+/// Determines if an on-chain price update is needed based on the latest Pyth price,
+/// the current on-chain price (if available), and the subscription's update criteria.
+#[tracing::instrument()]
+fn needs_update(
+    pyth_price: &Price,
+    chain_price_opt: Option<&Price>,
+    update_criteria: &UpdateCriteria,
+) -> bool {
+    // If there's no price currently on the chain for this feed, an update is always needed.
+    let chain_price = match chain_price_opt {
+        None => {
+            tracing::debug!("Update criteria met: No chain price available.");
+            return true;
+        }
+        Some(cp) => cp,
+    };
+
+    // 1. Heartbeat Check:
+    // Updates if `update_on_heartbeat` is enabled and the Pyth price is newer than or equal to
+    // the chain price plus `heartbeat_seconds`.
+    if update_criteria.update_on_heartbeat {
+        if pyth_price.publish_time
+            >= chain_price.publish_time + (update_criteria.heartbeat_seconds as i64)
```

Comment:
let's never use + , instead use `saturated_add` to make sure we never overflow (otherwise it might result in some security issues)